### ISSUE TITLE
restic reads too much data from the repository for an incremental backup

### ISSUE THREAD

**Issue by ckemper67** [restic/restic#582]
I noticed that my laptop (OSX MacBook Air) Chrome became unresponsive when running backups. I checked the ActivityMonitor and noticed a lot of network read activity attributed to restic. My backup seems to be mostly read bound.

My backups go to Google Cloud Storage (Google's S3 clone) and I also noticed that I got billed more for read access to my repository bucket than I would expect. I never performed a restore, so this was  bit surprising.

I decided to investigate and capture statistics during the backup. I suspect that this behavior is not isolated to GCS and can probably be reproduced using one of the other backends as well.

I am not sure if this is an inherent problem with restic's approach to handling the backups and the repository structure. It could be that it is reading a lot of packfiles to append some data, but in this case I would expect to see a lot more write activity as well since the updated packfiles need to be written out again. I would roughly expect that restic needs to read metadata

The stats before I killed the backup were: 655KB written,1.73GB read read of a 3.8GB directory tree.

## Output of `restic version`
restic compiled manually
compiled at unknown time with go1.6.2

I had built restic from a recent commit:
git log
commit 1c703e416134ddb542e24ff3917f7de83044dbf8
Merge: ca14942 ebd3723

## Expected behavior
restic reads data from backup repository at roughly the order of amount of metadata necessary to perform deduplication and writing the updated file content.

## Actual behavior

restic reads more data from the repository than is contained in the backup
du -h  ~/Documents shows 3.8G
My restic backup was reading 1.73GB from the repository at 32.18% done before I killed the backup

stats: 655KB written,1.73GB read,  32.18% progress

restic reported: scanned 2857 directories, 10957 files in 0:00

My repository contains 18 snapshots of this backup in case this matters

**Comment by fd0** [comment]
Hi, thanks for reporting this. At the moment, restic does not cache any data locally and keeps the state completely within the repository. The process for a new backup is roughly the following:
 * First, the index files are read. These describe which blobs are saved within which pack files in the repo, and (depending on the data) is rather small, in the order of several MiB per Index. 
 * Next, the snapshots are read, which are very small and only contains a bit of metadata fro each snapshot, in the order of a few KiB per snapshot.
 * When the backup starts, restic tries to find a previous snapshot for the same directories on the same host. If one is found, restic then loads all tree objects from the repository that are referenced by the parent snapshot and uses these to check which files have been modified. This can be disabled by running `restic backup -f`, which reads all files locally again, but will only write new data (-> deduplication).

This is the normal process, on the first backup there is no data read from the repo. For subsequent backups, the index files and the tree objects are read. Both should be rather small, maybe a few 10 MiB or so.

However, the library used for s3 access ([minio-go](https://github.com/minio/minio-go)) has a bug: When only a small range of a file in the repo should be read (tree objects are small), the whole flie is requested and downloaded (not completely, but much more data is requested than necessary): https://github.com/minio/minio-go/issues/484

TL;DR: Bug in the s3 library, plus no local cache at the moment.

**Comment by harshavardhana** [comment]
@fd0 would you mind validating these changes - https://github.com/minio/minio-go/pull/495 ?

**Comment by fd0** [comment]
This should be much better now, we've updated the minio-go library to include the fixes. Could you please give it a try again?

**Comment by ckemper67** [comment]
Yes it is much better now. Thanks for tracking this down.

This also fixed the ETA calculation. I never reported that the ETA was only counting up for me :-). I suspect with this fix the actual traffic to the backend is much more in line with the expectations of the ETA calculation.
